feat(forms): add error when adding control with dupe name to form

GrizzlyEnglish opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
    - Not sure on if this feature would need documentation other than api reference, if so please point me in the right direction


## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

Currently if a user adds controls to a form that share the same name, their could be unintended consequence due the models setting the controls by name in the form group and having no way to differentiate.

Issue Number: #13993 


## What is the new behavior?

When a control with a duplicate name is added to a form a runtime error is thrown to notify the user of the possibility of unintended consequences. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

I don't believe this would be a breaking change - since surely no one has existing apps with duplicate names.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

I tried avoiding adding any parameters or anything to existing methods in order to make as small of an impact as possible; but, due to the promises and the use case of ngFor renaming controls _as_ they are being deleted I had to add the changes to the control functions in order to determine if it was a rename or if the control was being added.

GrizzlyEnglish committed:
feat(forms): Added error when duplicate names are detected

This commits adds an error when a form is generated with more than one
control with the same name. Since there is no way to differentiate the
controls with the same name an error is now thrown.

This allows users to quickly see their mistake and fix it accordingly.

fixes #199339

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish commented:
My bad on the failed tests, not really sure what is happening 'cause if I run them, they pass.

<img width="527" alt="image" src="https://user-images.githubusercontent.com/12382017/182056796-8adfd0ba-8d9c-41f0-ac01-b111d4cbde48.png">

And I can forcefully make these tests fail by changing the expected header of the forms app

<img width="519" alt="image" src="https://user-images.githubusercontent.com/12382017/182057537-c9eb247a-9b2c-4634-a400-656700119312.png">

I am absolutely baffled currently, so my bad for having a PR with failed test 'cause like I said they work on my box :D. Probably due to environment more than anything.

atscott commented on packages/forms/src/directives/form_interface.ts:
```diff
@@ -28,8 +30,10 @@ export interface Form {
+   *     for dupes
    */
-  addControl(dir: NgControl): void;
+  addControl(dir: NgControl, changes: SimpleChanges): void;
```
@dylhunn At a glance, this change looks like it would be hugely breaking by adding a new required parameter to an existing function in the public API.

GrizzlyEnglish replied:
This is fair. I think I can refactor this in a way to add a function to the Form interface that will add the control with the validation and if the end user calls `addControl` it will assume the changes are null - cant image the end user adding a control for a name change.

AndrewKushnir replied:
I think we should move this logic to the `ngOnChanges` hook on the NgModel class instead and guard it with `ngDevMode` check (`if (ngDevMode) { ... the logic goes here ... }`), so that it's tree-shaken away in prod mode.

GrizzlyEnglish replied:
Thanks for the insight. Those are great ideas. I had wanted to put it in the `ngOnChanges`, but due to the add/remove of the controls being promises I wasn't sure of the best way to handle validating the current state of `controls`  is up to date. But I will look into. Hope it's cool I leave this PR open while I work through this refactor. Thanks!

AndrewKushnir replied:
@GrizzlyEnglish sure, no problem. I will mark your PR as "draft" for now to indicate that further changes are expected. Feel free to move it back to "ready for review" via GitHub UI once the implementation is updated.

Regarding the implementation itself, it looks like you may need to get a hold of the `_directives` private field (you can remove "private" modifier and add `/* @internal */` above it, so that it's accessible within the Forms package, but doesn't leak to the public API). Having that list available, the rest of the logic should work.

GrizzlyEnglish replied:
Awesome. Appreciate you @AndrewKushnir. I was afraid to change the private 😆

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

dylhunn commented:
I have reopened as requested in #13993. Would you like me to review this as-is? @GrizzlyEnglish

GrizzlyEnglish commented:
@dylhunn Yes please! Thanks!

dylhunn commented:
Great. My main concern on first glance is making this a non-breaking change. I'll get back to you soon with more specific comments.

dylhunn commented on packages/forms/src/directives/form_interface.ts:
```diff
@@ -6,6 +6,8 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {SimpleChanges} from '@angular/core';
```
This import is no longer needed, right?

GrizzlyEnglish replied:
Thanks nope, removed

dylhunn commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,7 +234,29 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+           * registered but given ngFor can produce duplicates *for a short time* we need to skip
+           * the validation when changes contain a previous value on the name, since the duplicate
+           * will be deleted. The promise is also necessary in order to add to the event stack prior
+           * to setup
```
> to setup

Nit: add period

GrizzlyEnglish replied:
Dropped using a promise in favor of ngZone, so this was removed.

dylhunn commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,7 +234,29 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+        if (this.formDirective) {
+          /**
+           * We need to check the directives for duplicates on the names when a control is
+           * registered but given ngFor can produce duplicates *for a short time* we need to skip
```
> registered but given ngFor

Nit: "registered. Given ngFor..."

GrizzlyEnglish replied:
Fixed

dylhunn reviewed (approved):
@GrizzlyEnglish This looks good to me, thanks for contributing! I have a couple super nitpicky comments below.

@AndrewKushnir It looks like all your comments have been addressed; mind giving this one more pass in case I missed something? ~Also do you know why pullapprove isn't requiring public-api reviews? It does touch a golden file.~ (Edit: it was just acting flaky, which is not new...)

reviewed-for: fw-forms, public-api

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -278,9 +301,9 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
   }
 
-  private _setUpControl(): void {
+  private _setUpControl(changes: SimpleChanges): void {
```
This change looks unnecessary now.

GrizzlyEnglish replied:
Removed

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,7 +234,29 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+           * to setup
+           */
+          Promise.resolve(null).then(() => {
+            if (changes != null && changes['name'] != null &&
+                changes['name'].previousValue == null) {
```
Actually, you can probably use `changes[name]?.isFirstChange()` here (if I understand correctly, the intention is to invoke the logic only the first time `name` is set).

GrizzlyEnglish replied:
Yeah `isFirstChange` much better didn't notice it before. Thanks got this updated.

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,7 +234,29 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+                if (!(d.valueAccessor instanceof RadioControlValueAccessor) &&
+                    !(this.valueAccessor instanceof RadioControlValueAccessor) &&
+                    d.name === this.name)
+                  throw duplicateFormNonRadioNames(this.name);
```
We've discussed this offline with @dylhunn and the proposal is to replace this `throw` with `console.warn` instead to avoid a breaking change.

GrizzlyEnglish replied:
Updated to a warn, and updated tests to match.

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,7 +234,29 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+          Promise.resolve(null).then(() => {
+            if (changes != null && changes['name'] != null &&
+                changes['name'].previousValue == null) {
+              this.formDirective._directives.forEach((d: NgModel) => {
```
```suggestion
              this.formDirective._directives.forEach((directive: NgModel) => {
```

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,7 +234,29 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+            if (changes != null && changes['name'] != null &&
+                changes['name'].previousValue == null) {
+              this.formDirective._directives.forEach((d: NgModel) => {
+                if (!(d.valueAccessor instanceof RadioControlValueAccessor) &&
```
Let's add a comment into the code to describe why we exclude the `RadioControlValueAccessor` from this check.

GrizzlyEnglish replied:
Added a comment in regards to the radio control needing duped names.

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,7 +234,29 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+           * will be deleted. The promise is also necessary in order to add to the event stack prior
+           * to setup
+           */
+          Promise.resolve(null).then(() => {
```
Note: we'd probably need to maintain a subscription as well and unsubscribe before subscribing again (and also unsubscribe in `ngOnDestroy`).

GrizzlyEnglish replied:
@AndrewKushnir  Thanks got this updated to utilize `ngZone` rather than a promise. With a empty subscription on creation and removal on detroy

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,7 +234,29 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+           * We need to check the directives for duplicates on the names when a control is
+           * registered but given ngFor can produce duplicates *for a short time* we need to skip
+           * the validation when changes contain a previous value on the name, since the duplicate
+           * will be deleted. The promise is also necessary in order to add to the event stack prior
+           * to setup
```
Not sure I fully understand this sentence. I think one of the main reasons we delay the check is to make sure that the data structures are updated and the update happens in `NgForm.addControl` in a Promise. The `NgZone.onStable` should help delay the check in this case as well (instead of adding a new Promise). it'd be great to update the comment here to explain this as well (so there is a sufficient context in the future on why we have a delayed operation here).

GrizzlyEnglish replied:
Thanks for the feedback. Yeah, my comment probably wasn't worded the best. Just wanted the dupe check to be placed on the end of the event loop in order to make sure it was after the completion of `NgForm.addControl`. Appreciate this and your other comments - will dig into understanding and implementing. Thanks!

AndrewKushnir reviewed (changes requested):
@GrizzlyEnglish thanks for updating the PR. I've left some comments. Please let me know if any additional information is needed. Thank you.

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,7 +234,29 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
       }
-      this._setUpControl();
+      if (ngDevMode) {
+        if (this.formDirective) {
```
Merged

GrizzlyEnglish commented:
@dylhunn @AndrewKushnir Apologies for the spam, and the delay. Got this updated based on feedback. Moved from using a promise to `ngZone` and an exception to console.warn. In doing so I had to update the tests as well. The only downside of this update was it was now triggering dupe names on everything; since, the `ngZone` triggers after and the directive exists in the collection. Just added a simple check to make sure it wasn't the same control.

Let me know if you see anything else. I appreciate the help!

dylhunn commented:
@GrizzlyEnglish Andrew has gone on vacation for the next few weeks. Let me re-review this and find a replacement second approver.

GrizzlyEnglish commented:
@dylhunn I am in no rush whatso ever if you choose to wait, I will work whatever bits needed later on. I just would like to try another ticket after this one is completed haha. Thanks!

dylhunn commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,6 +242,36 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+           * will be deleted.
+           */
+          if (changes['name']?.isFirstChange()) {
+            this.formDirective._directives.forEach((directive: NgModel) => {
```
@GrizzlyEnglish Your changes look great! 

This line is causing some code to break in the google3 presubmit. In particular: 

```
TypeError: Cannot read properties of undefined (reading 'forEach')

    at Object.next (third_party/javascript/angular2/rc/packages/forms/src/directives/ng_model.ts?l=258)
```

I haven't dug into this yet -- just wanted to let you know it seems that `_directives` is not always set.

GrizzlyEnglish replied:
Oh, my bad. I only focused on tests passing - and it seems I lack access to google3 - so it didn't occur to me to check the results. I am going to YOLO a null check on directives and just see if it fixes it hope that's good.

dylhunn replied:
It's probably good enough, although I would like to at least understand why. And indeed you don't have access to google3, so I'd need to do a bit of digging.

GrizzlyEnglish replied:
Ah understood. If there is anything I can do let me know. I can always roll back the null check. Appreciate all the help!

dylhunn commented:
Ahh, seems to be happening outside the Angular context. All the stack traces look like this:

```
          TypeError: Cannot read properties of undefined (reading 'forEach')

    at Object.next (forms/src/directives/ng_model.ts?l=258
    at SafeSubscriber.__tryOrUnsub (rxjs/src/internal/Subscriber.ts?l=268
    at SafeSubscriber.next (rxjs/src/internal/Subscriber.ts?l=210
    at Subscriber._next (rxjs/src/internal/Subscriber.ts?l=153
    at Subscriber.next (rxjs/src/internal/Subscriber.ts?l=113
    at Subject.next (rxjs/src/internal/Subject.ts?l=75
    at EventEmitter_.emit (core/src/event_emitter.ts?l=114
    at core/src/zone/ng_zone.ts?l=341
    at _ZoneDelegate.invoke (zone.js/lib/zone.ts?l=1145
    at FakeAsyncTestZoneSpec.onInvoke (zone.js/lib/zone-spec/fake-async-test.ts?l=674
    at ProxyZoneSpec.onInvoke (zone.js/lib/zone-spec/proxy.ts?l=134
    at _ZoneDelegate.invoke (zone.js/lib/zone.ts?l=1142
    at Zone.run (zone.js/lib/zone.ts?l=825
    at NgZone.runOutsideAngular (core/src/zone/ng_zone.ts?l=244
    at checkStable (core/src/zone/ng_zone.ts?l=341
```

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

dylhunn commented:
A status update on this:

- We need Andrew K's approval once he's back
- We should probably figure out why `_directives` is sometimes uninitialized
- A small number of internal presubmit failures, hopefully nothing too serious and we can sort them out.

Thanks again for the PR! :)

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,6 +242,36 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
           this.formDirective.removeControl({name: oldName, path: this._getPath(oldName)});
         }
       }
+      if (ngDevMode && this.formDirective) {
```
I'd propose moving the logic to a separate function for readability (to keep the `ngOnChanges` as simple as possible):
```suggestion
      if (ngDevMode) {
        _checkDuplicateModels(this, changes);
      }
```
Note: the `_checkDuplicateModels` is a standalone function (**not** within the NgModel class), so that it can be fully tree-shaken away.

GrizzlyEnglish replied:
Moved to standalone, took form check with it.

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,6 +242,36 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+           * the validation when changes contain a previous value on the name, since the duplicate
+           * will be deleted.
+           */
+          if (changes['name']?.isFirstChange()) {
```
Can we move this `if` outside of the subscribe callback?

GrizzlyEnglish replied:
Moved outside

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,6 +242,36 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+
+        this.dupeNameSubscription = this._ngZone.onStable.subscribe(() => {
+          /**
+           * We need to check the directives for duplicates on the names when a control is
+           * registered. Given ngFor can produce duplicates *for a short time* we need to skip
+           * the validation when changes contain a previous value on the name, since the duplicate
+           * will be deleted.
```
I believe this is the comment for the `this._ngZone.onStable.subscribe` call. If so, we should put the comment above the call.

GrizzlyEnglish replied:
I was attempting to explain why I was checking that name was in its first change

GrizzlyEnglish replied:
I think dropping the first sentence makes it a bit more concise

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -233,6 +242,36 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+              if ((!(directive.valueAccessor instanceof RadioControlValueAccessor) ||
+                   !(this.valueAccessor instanceof RadioControlValueAccessor)) &&
+                  directive !== this && directive.name === this.name)
+                console.warn(`
+                  Duplicate name "${
+                    this.name}" in form detected on non radio controls. This could lead to unintended results.
+                `);
```
```suggestion
                console.warn(`
                  `The "${this.name}" name is used for multiple \`ngModel\` bindings in this form. ` +
                  `This could lead to stale values in these bindings. ` +
                  `Make sure that elements with \`ngModel\` bindings in this form group ` +
                  `have unique "name" attribute values.`);
```

GrizzlyEnglish replied:
Updated. Thanks much better.

AndrewKushnir commented on packages/forms/test/template_integration_spec.ts:
```diff
@@ -2625,6 +2625,185 @@ import {NgModelCustomComp, NgModelCustomWrapper} from './value_accessor_integrat
+           const fixture = initTest(App);
+           fixture.detectChanges();
+           tick();
+           expect(console.warn).toHaveBeenCalled();
```
Let's also check the content of the message in the console, so that we don't accidentally treat another console.warn as a successful check here.

GrizzlyEnglish replied:
Updated each test to verify the warn contents

AndrewKushnir commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -216,6 +224,7 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
     this._setValidators(validators);
     this._setAsyncValidators(asyncValidators);
     this.valueAccessor = selectValueAccessor(this, valueAccessors);
+    this.dupeNameSubscription = null;
```
We can remove it (in favor of field initialization, see above).
```suggestion
```

AndrewKushnir reviewed (commented):
Looks great, just some additional feedback.

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -157,6 +159,12 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+   * A subscription to the duplicate name check. Allow null as initial value.
+   * @nodoc
+   */
+  private dupeNameSubscription: Subscription|null;
```
Updated to init as null here rather than constructor

GrizzlyEnglish commented:
@dylhunn @AndrewKushnir Apologies for taking so long on this. Life happened. Question for you two: I rebased like mentioned by circle ci - and now it thinks my latest commit is not a tree. Thoughts on this?

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish commented:
Ok, rebased again and now that is all fine - but it seems I have gone over the limit

GrizzlyEnglish commented:
@dylhunn Sorry to be a pain on this. I am not sure how to fix this. My most recent commit [a14f30838a2606d73098635ed3f30828fd25a640 ](https://github.com/angular/angular/pull/46994/commits/a14f30838a2606d73098635ed3f30828fd25a640) is saying that it is too large; but it is the required symbols.

jessicajaniuk commented:
@GrizzlyEnglish You should be able to look at the test failure and update the golden file with the new size. The file is in the root of the repo at [goldens/size-tracking/integration-payloads.json](https://github.com/angular/angular/blob/main/goldens/size-tracking/integration-payloads.json). Update the block for `forms > uncompressed` with the new values. Once you do that, we'll get pinged for size tracking reviews.

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish commented:
@jessicajaniuk Thank you!

jessicajaniuk commented:
@GrizzlyEnglish looks like your fixup commit updated the size to the expected value and not the actual value. The failing size test shows the main uncompressed for forms actual value as 158276. Setting it to that should get it passing.

GrizzlyEnglish committed:
fixup! feat(forms): Added error when duplicate names are detected

GrizzlyEnglish commented:
@jessicajaniuk Ah my bad. Working and working on this not a good combo :D

dylhunn commented:
I or Andrew will look at this shortly, but for now let's target v15.1.

JeanMeche commented:
Do you mind rebasing your PR and fix the conflicts ?

thePunderWoman commented:
Since we've seen no movement on this PR, I'm going to close this. If you'd like to see these changes land, feel free to open a new PR.
